### PR TITLE
tests: adding a test for tree view

### LINKED ISSUE #25: Typing issues in python 3.8.6 (broken?)
Quick reproduction:

```
docker run -it --rm python:3.8 bash -c 'pip install --quiet uproot-browser; curl -sLO https://github.com/scikit-hep/scikit-hep-testdata/raw/main/src/skhep_testdata/data/uproot-issue213.root; uproot-browser browse uproot-issue213.root'
```

Would have expect this to work out of the box, but seems to be broken with this (very unhelpful) stack trace:

```
Traceback (most recent call last):
  File "/usr/local/bin/uproot-browser", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/uproot_browser/__main__.py", line 74, in browse
    import uproot_browser.tui  # pylint: disable=import-outside-toplevel
  File "/usr/local/lib/python3.8/site-packages/uproot_browser/tui.py", line 14, in <module>
    from .tree_view import TreeView, UprootClick
  File "/usr/local/lib/python3.8/site-packages/uproot_browser/tree_view.py", line 15, in <module>
    from .tree import UprootItem
  File "/usr/local/lib/python3.8/site-packages/uproot_browser/tree.py", line 82, in <module>
    def _process_item_tfile(
  File "/usr/local/lib/python3.8/functools.py", line 858, in register
    argname, cls = next(iter(get_type_hints(func).items()))
  File "/usr/local/lib/python3.8/typing.py", line 1264, in get_type_hints
    value = _eval_type(value, globalns, localns)
  File "/usr/local/lib/python3.8/typing.py", line 270, in _eval_type
    return t._evaluate(globalns, localns)
  File "/usr/local/lib/python3.8/typing.py", line 518, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
TypeError: 'type' object is not subscriptable
```

### PR DESCRIPTION
**Pull request by henryiii**
Closes #25.

### CODE REVIEW

**Comment by jpivarski** [comment]
No, I don't. Is the file uproot-issue213.root?

**Comment by jpivarski** [comment]
No, it's uproot-Event.root. This branch:

```python
>>> tree = uproot.open(skhep_testdata.data_path("uproot-Event.root") + ":T")
>>> tree["event/fTriggerBits/fTriggerBits.fAllBits"].interpretation
AsJagged(AsDtype('uint8'), header_bytes=1)
>>> tree["event/fTriggerBits/fTriggerBits.fAllBits"].interpretation.typename
'uint8_t[]'
```

**Comment by henryiii** [comment]
Yes, for me that gives:

```python
>>> tree = uproot.open(skhep_testdata.data_path("uproot-Event.root") + ":T")
>>> tree["event/fTriggerBits/fTriggerBits.fAllBits"].interpretation
AsJagged(AsDtype('uint8'), header_bytes=1)
>>> tree["event/fTriggerBits/fTriggerBits.fAllBits"].interpretation.typename
'uint8_t*'
```

**Comment by henryiii** [comment]
It's technically the same, but an irritating difference. I could search&replace it out of the output if needed.

**Comment by jpivarski** [comment]
Hmmm. I'm following the path it takes through [uproot.interpretation.identify.interpretation_of](https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L294).

  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L363
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L380
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L63
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L96
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L31
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L60 because `fType` is `51`
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L425
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L432
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L446
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L462
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L470 and `model_cls` is `AsArray(False, True, dtype('uint8'), ())`, which has a `typename` of `"uint8_t[]"`, so let's descend into `parse_typename`
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L1074
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L1101 and `tokens` is a list of match objects corresponding to `"unsigned"`, `"char*"`
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L563
  * https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L617

And that's how we get the final interpretation on Linux. Wait a minute: you're saying you see the same interpretation, just a different typename? Let me look into that...

**Comment by henryiii** [comment]
`int*` and `int[]` are the same, so if it’s just (expected) differences in the underlying implementation, I can just always replace one with the other in the test.

**Comment by jpivarski** [comment]
Back to where we have a `AsArray(False, True, dtype('uint8'), ())` on https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L470, the next step is that this goes into AsObjects, which then gets simplified:

https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L489-L493

  * simplify: https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/objects.py#L239
  * enters https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/objects.py#L247, but it `CannotBeStrided`
  * it's not `AsString`, but it is `AsArray`: https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/objects.py#L273
  * it has one `header_bytes` (due to the `speedbump`) and the `values` are a NumPy dtype, so we get to https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/objects.py#L288 That's how the final interpretation becomes AsJagged. The `typename` in the output comes from the `self._model.typename`, which is `"uint8_t[]"` on Linux.

So how did the `self._model` get that `typename`? Following that path now...

**Comment by jpivarski** [comment]
The reason I'm deep-diving into this is that there shouldn't be any platform-dependent difference. It's disturbing me. The ROOT file is the same set of bytes, and if there's a platform-dependency in how we grab those bytes, I want to know about it.

(Like all of Window's `int32` differences. Knowing about them prevents future bugs.)

**Comment by jpivarski** [comment]
The `model`, which has a `typename` that gets passed to the simplified final interpretation, is created by evaluating the string made here:

https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L620-L622

which is exactly

```python
uproot.containers.AsArray(False, True, numpy.dtype("u1"))
```

That should be entirely platform-independent: the only variable going into making that string is `header`, which must be `"True"` because you see the same number of header bytes (`1`, not `0`).

If I evaluate

```python
>>> import uproot, numpy
>>> uproot.containers.AsArray(False, True, numpy.dtype("u1")).typename
'uint8_t[]'
```

on Linux, I get the brackets, as you can see above. On any platform, this has to produce brackets, not a `*`, because the `AsArray.typename` is defined like this:

https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/containers.py#L628-L631

Before entering `simplify` ([this block](https://github.com/scikit-hep/uproot4/blob/8e3ab33c2d137079db53cd2c21b4c3ac7e1dcffc/src/uproot/interpretation/identify.py#L489-L493)), is your `model_cls` not an AsArray? Could it possibly be an AsPointer? That's the only type that has a `typename` with a `"*"` in it.

**Comment by jpivarski** [comment]
Maybe if regular expressions act a little differently:

```python
>>> import re
>>> _tokenize_typename_pattern = re.compile(
...     r"(\b([A-Za-z_0-9]+)(\s*::\s*[A-Za-z_][A-Za-z_0-9]*)*\b(\s*\*)*|<|>|,)"
... )
>>> list(_tokenize_typename_pattern.finditer("unsigned char*"))
[
    <re.Match object; span=(0, 8), match='unsigned'>,
    <re.Match object; span=(9, 14), match='char*'>,
]
```

If the tokenizer results in three tokens, corresponding to `"unsigned"`, `"char"`, and `"*"`, rather than the two above, that would make the model resolve into an AsPointer and a dtype, rather than an AsArray and a dtype. If that's the case, I want to be sure that we use regular expressions in a platform-independent way in the future. (The intended behavior is what Linux is doing, but that's because it was developed on Linux.)

**Comment by henryiii** [comment]
```python
>>> import uproot, numpy
>>> uproot.containers.AsArray(False, True, numpy.dtype("u1")).typename
'uint8_t*'
```

**Comment by henryiii** [comment]
Ahh, I was using uproot 4.0.11. Upgrading to 4.2.2 gives the `'uint8_t[]'` answer!

**Comment by jpivarski** [comment]
Oh, okay. (I'm breathing a sigh of relief.) 4.0.11 may be very different from 4.2.2; there's a lot of versions in between. I don't know what change it might have been (probably not the parsing, I'd guess), but not something that I need to look into.

File: pyproject.toml [hunk 1]
```diff
@@ -75,7 +76,10 @@ uproot-browser = "uproot_browser.__main__:main"
 minversion = "6.0"
 addopts = ["-ra", "--showlocals", "--strict-markers", "--strict-config"]
 xfail_strict = true
-filterwarnings = ["error"]
+filterwarnings = [
+  "error",
+  "ignore:can't resolve package from __spec__ or __package__, falling back on __name__ and __path__:ImportWarning",  # PyPy NumPy
```
Comment (henryiii) [inline pyproject.toml:81]: @mattip, this is an odd warning from NumPy PyPy that gets produced if you enable all warnings (`filterwarnings=["error"]` for pytest, or `-Werror`)

https://github.com/henryiii/uproot-browser/runs/5586825626?check_suite_focus=true

Comment (mattip) [reply pyproject.toml:81]: I have seen that but I have no idea what it means.

**Pull request merged**